demos/pkcs11/pkcs11_demo_mechanisms_and_digests/pkcs11_demo_mechanisms_and_digests.c

```diff
@@ -0,0 +1,283 @@
+    }
+
+    assert( MechanismInfo.ulMaxKeySize >= pkcs11ECDSA_P256_KEY_BITS );
+    assert( MechanismInfo.ulMinKeySize <= pkcs11ECDSA_P256_KEY_BITS );
+
+    /************************** Digest Capabilities **************************/
+    xResult = pxFunctionList->C_GetMechanismInfo( pxSlotId[ 0 ],
+                                                  CKM_SHA256,
+                                                  &MechanismInfo );
+    assert( CKR_OK == xResult );
+
+    if( 0 != ( CKF_DIGEST & MechanismInfo.flags ) )
+    {
+        LogInfo( ( "The Cryptoki library supports the " \
+                   "SHA-256 algorithm." ) );
+    }
+    else
+    {
+        LogInfo( ( "The Cryptoki library doesn't support the " \
+                   "SHA-256 algorithm." ) );
+    }
+
+    /***************************** Buffer Digest *****************************/
+    /* Hash with SHA256 mechanism. */
+    xDigestMechanism.mechanism = CKM_SHA256;
+
+    /* Initializes the digest operation and sets what mechanism will be used
+     * for the digest. */
+    xResult = pxFunctionList->C_DigestInit( hSession,
+                                            &xDigestMechanism );
+    assert( CKR_OK == xResult );
+
+
+    /* Pass a pointer to the buffer of bytes to be hashed, and it's size. */
+    xResult = pxFunctionList->C_DigestUpdate( hSession,
+                                              pxKnownMessage,
+                                              /* Strip NULL Terminator. */
+                                              sizeof( pxKnownMessage ) - 1 );
+    assert( CKR_OK == xResult );
+
+    /* Retrieve the digest buffer. Since the mechanism is a SHA-256 algorithm,
+     * the size will always be 32 bytes. If the size cannot be known ahead of time,
+     * a NULL value to the second parameter pDigest, will set the third parameter,
+     * pulDigestLen to the number of required bytes. */
+    xResult = pxFunctionList->C_DigestFinal( hSession,
+                                             xDigestResult,
+                                             &ulDigestLength );
+    assert( CKR_OK == xResult );
+
+    /* This will now print out the digest of the known message. You can compare
+     * the hash generated by the Cryptoki library in a UNIX shell by using the
+     * command '$ echo -n "{pxKnownMessage}" | shasum -a 256'
+     * this command should generate the same hash. */
+    LogInfo( ( "Known message: %s", ( char * ) pxKnownMessage ) );
+
+    pcFormatPtr = &xHashFormatBuffer[ 0 ];
+
+    for( ulIndex = 0; ulIndex < ulDigestLength; ulIndex++ )
+    {
+        pcFormatPtr += sprintf( ( char * ) pcFormatPtr, "%x", xDigestResult[ ulIndex ] );
+    }
+
+    LogInfo( ( "Hash of known message using SHA256: %s.", xHashFormatBuffer ) );
+    LogInfo( ( "Finished PKCS #11 Mechanisms and Digest Demo." ) );
+
+    /* Avoid compiler warnings if asserts are disabled. */
+    ( void ) xResult;
+
+    vEnd( hSession, pxSlotId );
```

Comment:
Does one of the PKCS libraries expose this function?
If relevant in the demo, a comment about it would be useful

Reply:
This is for convenience to not repeat code explained in the first demo